---
title: "COVID proteomics: neutrophils"
author: "Marek Gierlinski"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme:
      bootswatch: journal
      primary: "#2fa4e7"
editor_options: 
  markdown: 
    wrap: 72
---

```{css css_options, echo=FALSE}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll_box() ! important; 
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

```{r options, echo=FALSE}
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  echo = FALSE,
  cache = TRUE,
  autodep = TRUE
)
options(
  width = 100,
  knitr.table.format = "html"
)
```

```{r libraries, cache=FALSE}
library(targets)
library(tidyverse)
library(flextable)
library(UpSetR)
library(DT)
```

```{r functions}
myDT <- function(d) {
  datatable(d, class = "table-condensed table-striped table-bordered", style = "bootstrap") %>% 
    formatStyle(columns = colnames(d), fontSize = '90%')
}
N <- function(n) prettyNum(n, big.mark = ",")
```

Collaborators: Merete Long, James Chalmers

#  {.tabset}

## Proposal

### STOP-COVID19: SUPERIORITY TRIAL OF PROTEASE INHIBITION IN COVID-19

Multi-centre prospective, randomized placebo-controlled trial in
patients presenting to hospital with PCR confirmed COVID-19 (SARS-CoV-2
infection). Planned Sample Size was 300 across the UK. Treatment
duration was 28 days, with 29 days follow-up.

#### Study rationale and hypothesis

COVID-19 can cause substantial morbidity and mortality. Mortality in
acute respiratory distress syndrome (ARDS) correlates directly with the
extent of neutrophilia in the lung. Both human clinical data and murine
studies demonstrate a key role for neutrophils in ARDS. Neutrophil
proteases and particularly neutrophil elastase are believed to be
central to the neutrophil induced lung damage.

Neutrophil elastase, proteinase-3 and cathepsin-G are activated during
neutrophil maturation in the bone marrow through dipeptidyl peptidase 1
(DPP1; also known as cathepsin C), which removes the N-terminal
dipeptide sequence of neutrophil serine proteases allowing active
enzymes to be packaged into granules prior to release of neutrophils
into the circulation. Brensocatib (INS1007, formerly AZD7986) is an
orally delivered selective, competitive, and reversible inhibitor of
DPP1. Brensocatib has been shown to inhibit neutrophil serine protease
activity in blood in both animal models and healthy volunteers and in
the airways of patients with the chronic respiratory disease
Bronchiectasis.

**We hypothesised that Brensocatib, by blocking damaging neutrophil
proteases, would reduce the incidence of acute lung injury and acute
respiratory distress syndrome (ARDS) in patients with COVID-19, thereby
resulting in improved clinical outcomes.**

### Study overview

In total 144 participants were enrolled in Dundee, and 17 enrolled in
Sheffield, both of these sites obtained longitudinal research blood
samples from participants whilst hospitalised, for exploratory analyses
to better understand Brensocatib mechanisms of action. Participants were
also invited to return for sampling a day 29 if discharged, resulting in
larger sample numbers at day 29 than at intermediate timepoints.
Isolated peripheral blood neutrophils across various timepoints have
been processed and analysed with LC-MS, utilising Spectronaut software
for protein identification.

### Major questions for initial proteomic data analysis

-   Does treatment with Brensocatib alter the COVID-19 peripheral blood
    neutrophil proteome across the study period, compared with Placebo
    treatment?

-   Do factors such as age, gender, or time from symptom onset influence
    responses to Brensocatib within the neutrophil proteome?

-   If possible, the above would be performed in the study population
    who completed the trial (completed=1), and then would also be
    performed subsequently as an intention-to-treat analysis, including
    all participants (both completed=1 and completed=0)

### Data Analysis Group Quotation Reference: 150422-JC03

Project Title: Neutrophil/PBMC proteomic analysis

1.  PI Details

Name Prof. James Chalmers Division Molecular and Clinical Medicine
School School of Medicine Email
[j.chalmers\@dundee.ac.uk](mailto:j.chalmers@dundee.ac.uk){.email}

2.  Background

From May 2020 we carried out a trial of dipeptidyl
peptidase-1/Cathepsin-C inhibitor in hospitalised patients with
COVID-19. In total enrolled 144 participants in Dundee. We have data
from around 250 neutrophil samples across various timepoints which have
been processed and analysed with LC-MS, subsequently utilising
Spectronaut software for protein identification. Comparisons of drug vs.
placebo groups at 4-5 timepoints is requested, with possible
sub-analyses for e.g. age and gender, time from symptom onset. Batch
correction across 3-5 batches would also be needed, as significant batch
effects were found in other analyses of the proteomics data. If possible
we would also do the same thing for the PBMC samples later, however the
neutrophil data would be a first priority.

3.  Description of Proposed Work

This is a large and complex experiment, so would be better to cost in
stages. The first stage would include:

-   overview of the data

-   quality control

-   multidimensional analysis: finding important and non-important
    factors, dealing with batch effects

-   differential abundance on selected conditions (e.g. day 29 vs day 1,
    drug vs placebo at a given time point)

-   multivariate differential abundance on time + treatment + batch,
    etc. This might not work if data are not "clean", e.g., if timing
    behaviour is complex (i.e. not a simple trend)

-   simple timing analysis, identifying proteins consistently changing
    up or down

-   providing a simple interactive tool to look at the above results

## Data

### Spectronaut

Raw data were analysed with
[Spectronaut](https://biognosys.com/software/spectronaut). I have
received two files: one is the original output from Spectronaut (a TSV
file in long format), the other is a file processed by collaborators (a
CSV file in wide format). I did initial analysis using the processed
file (I got it first), but then switched to the original file.

The file I used is a protein quantification file and it contains the
following columns:

```{r spectronaut_columns}
tar_read(spectronaut_columns) %>% 
  flextable()
```

We need file name (corresponding to the sample name), protein
accessions, descriptions and quantity - raw protein intensity. All
q-values are small (the largest q-value is
`r signif(tar_read(max_qvalue), 2)`).

### Normalisation

Raw data are normalised to the median in each sample.

## Experiment

### Batches

We have `r tar_read(n_samples)` samples. There are two treatments (drug
and placebo), several time points and batches. Here is a summary of the
sample numbers:

```{r fig_counts, fig.width=6, fig.height=4}
tar_read(fig_counts_batch) 
```

We see that batch 1 and 2 are very small (we are going to ignore these data when taking batch effects into account) and days 4 and 15 have little data available. More details in tabular form:

#### Treatment and day

```{r meta_day}
tar_read(metadata) %>%
  group_by(treatment, day) %>% 
  tally() %>% 
  flextable() %>% 
  theme_box() %>% 
  merge_v(j = 1)
```

#### Batch and day

```{r meta_batch}
tar_read(metadata) %>%
  group_by(batch, day) %>% 
  tally() %>% 
  flextable() %>% 
  theme_box() %>% 
  merge_v(j = 1)
```

### MS runs

```{r fig_counts_run, fig.width=6, fig.height=4}
tar_read(fig_counts_run) 
```

MS runs are subsets of batches:

```{r meta_batch_run}
tar_read(metadata) %>%
  group_by(batch, run_index) %>% 
  tally() %>% 
  flextable() %>% 
  theme_box() %>% 
  merge_v(j = 1)
```



### Full table of the experimental design

Please note that sample naming in this report (`sample`) differs from
the file/sample naming in the Spectronaut file (`raw_sample`). New
sample name consists of the following:

-   treatment (`P` for placebo, `D` for drug)
-   batch number (1-5)
-   day (01, 04, 08, 15, 29)
-   'replicate' number (1 .. n, within given treatment/batch/day)

```{r meta}
m <- tar_read(metadata) %>% 
  select(sample, raw_sample, batch, run_index, participant_id, treatment, day, age, sex, completion)
DT::datatable(m) %>%  formatStyle(columns = colnames(m), fontSize = '80%')
```

## Participants

### Table of participants

```{r participants}
DT::datatable(tar_read(participants)) 
```

### Distribution of attributes

```{r fig_participant_stats, fig.width=11, fig.height=3.5}
tar_read(fig_participat_stats) 
```

## Overview

### Protein detection

Not all proteins are detected across all `r tar_read(n_samples)` samples. Figures below show

 - in how many samples each protein was detected, starting from the highest number
 - how many proteins are detected in each sample, starting from the highest number
 

```{r detection, fig.width=10, fig.height=4}
tar_read(fig_detection)
```

There is a dramatic drop at the end of the right panel, which is due to one sample  with very few proteins detected. This sample is removed from further analysis. Here is information about the bad sample:


```{r bad_samples}
m <- tar_read(metadata) %>% 
  filter(bad) %>% 
  select(sample, raw_sample, batch, participant_id, treatment, day, age, sex, completion)
flextable(m) %>% 
  width(j = 1:2, width = 1.3)
```

### Abundance distribution

The plot shows distribution of raw and normalised protein abundance:

```{r fig_abu_dist, out.width="60%"}
knitr::include_graphics("../fig/kernels.png")
```

### Clustering

OK, this is a monstrous figure, but allows to see all the samples clustered together. Colours of sample names represent batches. We can see that batch effects dominate this experiment. Hierarchical clustering was calculated based on logarithm of normalised intensities, with Euclidean distance, using complete linkage method.

```{r fig_clustering, fig.width=8, fig.height=40}
tar_read(fig_clustering)
```

### Correlation matrix

This figure shows Pearson's correlation coefficient for all pairs of samples. It is a bit messy. The bad sample is clearly visible as a dark row and column. The figure is based on logarithm of normalised intensities.

```{r cormat, fig.width=10, fig.height=10}
tar_read(fig_cormat) 
```

### PCA

The figure is based on logarithm of normalised intensities. The bad replicate stands out.

```{r pca, fig.width=8, fig.height=8}
tar_read(fig_pca)
```

### UMAP

UMAP is dimensionality reduction method, an alternative to PCA. It retains structure of high-dimensional data when embedded in two dimensions.

```{r umap, fig.width=8, fig.height=8}
tar_read(fig_umap) 
```

## Differential abundance

### Description

#### From limma R vignette

For differential expression we use `limma`. It used an empirical Bayes
method to squeeze the peptide-wise residual variances towards a common
value (or towards a global trend) (Smyth, 2004; Phipson et al, 2016).
The degrees of freedom for the individual variances are increased to
reflect the extra information gained from the empirical Bayes
moderation, resulting in increased statistical power to detect
differential expression.

For more information see [Ritchie et al.
2015](https://academic.oup.com/nar/article/43/7/e47/2414268).

#### In simpler words

`limma` uses a t-test, which is moderated by borrowing data across all
peptides. A global variance model is built and moderation squeezes
variances towards this global trend. Extreme variances (either large or
small) become less extreme, which makes the test more robust to random
outliers.

### Full model

First, we analyse all data using a 'full' model taking into account all potential effects:

`~ treatment + day + batch + age_group + sex`

The differential abundance is calculated with respect to the baseline:

 - treatment = placebo
 - day = 1
 - batch = 3
 - age_group = <50
 - sex = F
 
In this model, and models below, bad replicate and data from batch 1 and 2 are ignored.

The figure below shows MA plots for each of the coefficients. Each panel corresponds to one value of one coefficient, showing differential abundance with respect to the baseline.

```{r fig_ma_full, fig.width=10, fig.height=10}
tar_read(fig_ma_full) 
```

Black points indicate proteins with significant -- false discovery rate (FDR) < 0.01 -- differential abundance. Note that with dozens of samples even small changes become statistically significant, so we need to pay attention to the effect size (log fold change).

The biggest effect comes from batch differences and time points. The drug effect is week, though it shows a few differentially abundant (DA) proteins, these might be flukes. We are going to look at drug effect in details later. Similarly, other factors (age and sex) have very week effect on DA.


### Batch effect

Next, we look at the difference between batches 3, 4 and 5. For this, a simple model with only batch as effect was created and three contrasts calculated:

```{r fig_da_batch, fig.width=8, fig.height=3}
tar_read(fig_ma_batch)
```

Somehow batches 4 and 5 are the most different, but statistical significance can be misleading here due to a large number of samples and a complex design.

A couple of examples to demonstrate the strength of batch effect. MS run indeces are also included.

```{r fig_batch_examples, fig.width=9, fig.height=5}
tar_read(fig_batch_examples)
```

Trypsin is an extreme example where batches 4 and 5 differ by almost 3 orders of magnitude.

#### Gene set enrichment for batch effect

```{r gse_batch5}
tar_load(gse_batch5)
map_dfr(names(gse_batch5), function(ont) {
  gse_batch5[[ont]] %>% 
    filter(padj < 0.01) %>% 
    add_column(ontology = ont, .before = 1) %>% 
    mutate(across(c(pval, padj, NES), ~signif(.x, 3))) %>% 
    select(ontology, term, term_name, padj, NES, size, contrast) %>% 
    mutate(ontology = recode(ontology, go = "GO", re = "Reactome", kg = "KEGG"))
}) %>% 
  DT::datatable()
```



### Day 1

We select day 1 data only and consider a model

`~ treatment + batch + age_group + sex`

```{r fig_ma_day1, fig.width=8, fig.height=8}
tar_read(fig_ma_day1) 
```

Not surprisingly there is no drug effect on day 1.


### Day 29

Now the same for day 29. Note that there is no batch 3 in day 29 data.

```{r fig_ma_day29, fig.width=8, fig.height=5.5}
tar_read(fig_ma_day29) 
```

Now a lot of proteins are 'significant' in drug effect, but the fold change is not any different from day 1. We need to be careful interpreting these results. Again, they are completely dominated by batch effects.


### Day 29 and batch effects

There are two batches within day 29: batch 4 and batch 5. As demonstrated above, the batch effect is strong. Here, two data subsets are considered, for batch 4 and batch 5 separately (both for day 29). The model is 

`~ treatment + age_group + sex`

```{r fig_day29_4, fig.width=8, fig.height=5.5}
tar_read(fig_ma_day29_b4) + ggtitle("Batch 4")
```


```{r fig_day29_5, fig.width=6, fig.height=5.5}
tar_read(fig_ma_day29_b5) + ggtitle("Batch 5")
```

The next figure shows overlap for DA proteins (FDR < 0.01) for day 29: batch 4, batch 5 and both:

```{r day29_upset}
upset(fromList(tar_read(ups_day29))) 
```

The only proteins in agreement between three selections are:

```{r day29_best}
tar_read(dal_day29_treatment_best) %>% 
  select(-protein_names) %>% 
  flextable() %>% 
  width(2, 2) %>% 
  width(3, 3)
```

Here are abundances from these proteins:

```{r fig_day29_best, fig.width=10, fig.height=4}
tar_read(fig_day29_best)
```

### The outlier

There is one persistent outlier keeping high levels of AZU1 and CTSG on day 29. It is the same participant, ID 1132, one of the youngest in the cohort:

```{r part1132}
tar_read(metadata) %>% 
  filter(participant_id == 1132) %>% 
  select(participant_id, age, sex, time_from_symptoms, completion) %>% 
  distinct() %>% 
  flextable()
```

These are the same five proteins for this participant 1132:

```{r fig_day29_best_1132, fig.width=10, fig.height=4}
tar_read(fig_day29_best_part1132)
```


## Per participant DA


Protein measurements at different time points are not independent, as they come from the same individuals. The plot below shows which participants gave samples at which days:

```{r fig_participants, fig.width=10, fig.height=5}
tar_read(fig_participant_1_29)
```

Black lines indicate participants delivering samples both on day 1 and 29. We can use these data to link day 1 and 29 data and make a paired model. For this, a log-ratio of intensity at day 29 to day 1 is calculated for each participant and each protein. If there is no change in protein abundance between days 1 and 29, the log-ratio should be close to 0. We perform a one-sample test for each protein with the null hypothesis of 0. We use `limma` to control variance between proteins.

This should give us differential abundance in time (days 1 and 29), corrected for interdependence (participants).

```{r dd_enrichment}
tar_load(gse_dd)
tb <- map_dfr(names(gse_dd), function(ont) {
  gse_dd[[ont]] %>% 
    filter(padj < 0.01) %>% 
    add_column(ontology = ont, .before = 1) %>% 
    mutate(across(c(pval, padj, NES), ~signif(.x, 3))) %>% 
    select(ontology, term, term_name, padj, NES, size, contrast) %>% 
    mutate(ontology = recode(ontology, go = "GO", re = "Reactome", kg = "KEGG"))
})
```

### Volcano plots

```{r fig_volcano_dl, fig.width=7, fig.height=3}
tar_read(fig_volcano_dl)
```


### Placebo set enrichment


```{r fig_dd_placebo, fig.width=4, fig.height=3}
DT::datatable(tb %>% filter(contrast == "placebo") %>% select(-contrast))
```

### Drug set_enrichment


```{r fig_dd_drug, fig.width=4, fig.height=3}
DT::datatable(tb %>% filter(contrast == "drug") %>% select(-contrast))
```

### Drug vs placebo

We can compare a day 1-to-29 change between drug and placebo.

```{r fig_volcano_dd_drug_vs_placebo, fig.width=8, fig.height=5.5}
tar_read(fig_volcano_dl_drug_vs_placebo)
```

Here are the significant proteins:

```{r dd_drug_vs_placebo}
tar_load(dal_dl_drug_vs_placebo_proteins)
DT::datatable(dal_dl_drug_vs_placebo_proteins %>% select(-protein_names)) 
```



```{r fig_dd_drug_vs_placebo, fig.width=10, fig.height=8}
tar_read(fig_dl_drug_vs_placebo) 
```


## Gene set enrichment

Gene set enrichment takes genes/proteins ordered by log-fold-change (ignoring p-values and significance) and looks for functional terms that are skewed towards small or large logFC. For example, GO-term "oxygen carrier activity" in the figure below is assigned to several proteins (haemoglobin and histone related) that all are down-regulated in age group over 80. None of these proteins is significant in simple differential abundance, but as a functional group they become significant, as they all are down-regulated.

Here we use fold changes from day 29 selection, for treatment and age contrasts (batch excluded). The baseline is treatment = placebo and age_group = '<50'. Contrast names are at the top of each panel. For example `age_group>80` means comparison between `>80` and the baseline `<50`.

The figure below shows volcano plots for various contrasts and functional terms, with term-annotated proteins highlighted.

```{r, out.width="100%"}
knitr::include_graphics("../fig/gse_examples.png") 
```

Below is the full result of gene-set-enrichment procedure (excluding batch effects). The FDR level was set to 0.05. NES stands for normalised effect size and tells you how far the group of proteins annotated with this term is skewed in logFC. Small and large NES means proteins are predominantly down- or up-regulated, respectively. The last column contains up to 10 top most changing genes (in the same direction as NES).

```{r gse_results}
tar_read(gse_da_day29_sig) %>% 
  mutate(ontology = recode(ontology, go = "GO", re = "Reactome", kg = "KEGG")) %>% 
  DT::datatable()
```

## Resources

-   [Interactive data explorer](https://shiny.compbio.dundee.ac.uk/private/marek_covid_prot_neutrophils/da_day29)
-   [Download DE results]()

## Session info

```{r session_info}
sessionInfo()
```
